crates/sui-core/src/narwhal_manager/mod.rs

```diff
@@ -54,128 +33,127 @@ pub struct NarwhalConfiguration<TxValidator: TransactionValidator> {
-            registry_service: config.registry_service.clone(),
-        };
-
-        start_narwhal(
-            config_copy,
-            committee,
-            shared_worker_cache,
-            execution_state,
-            &primary_node,
-            &worker_nodes,
-        )
-        .await;
-
-        // Wait for instruction to stop the instance of narwhal
-        match tr_stop.recv().await {
-            Some(_) => shutdown_narwhal(&primary_node, &worker_nodes).await,
-            None => break,
-        };
+        // Create a new store
+        let mut store_path = self.storage_base_path.clone();
+        store_path.push(format!("epoch{}", committee.epoch()));
+        let store = NodeStorage::reopen(store_path);
+
+        let name = self.primary_keypair.public().clone();
+
+        let now = Instant::now();
+        tracing::info!("Starting up Narwhal for epoch {}", committee.epoch());
+
+        // start primary
+        self.primary_node
+            .start(
+                self.primary_keypair.copy(),
+                self.network_keypair.copy(),
+                Arc::new(ArcSwap::new(committee.clone())),
+                shared_worker_cache.clone(),
+                &store,
+                execution_state,
+            )
+            .await
+            .expect("Unable to start Narwhal Primary");
+
+        // Start Narwhal Workers with configuration
+        // Copy the config for this iteration of the loop
+        let id_keypair_copy = self
+            .worker_ids_and_keypairs
+            .iter()
+            .map(|(id, keypair)| (*id, keypair.copy()))
+            .collect();
+
+        self.worker_nodes
+            .start(
+                name,
+                id_keypair_copy,
+                Arc::new(ArcSwap::new(committee.clone())),
+                shared_worker_cache,
+                &store,
+                self.tx_validator.clone(),
+            )
+            .await
+            .expect("Unable to start Narwhal Worker");
+
+        tracing::info!(
+            "Starting up Narwhal for epoch {} is complete - took {} seconds",
+            committee.epoch(),
+            now.elapsed().as_secs_f64()
+        );
+
+        *running = true
     }
-}
 
-#[instrument(level = "info", skip_all)]
-async fn shutdown_narwhal(primary_node: &PrimaryNode, worker_nodes: &WorkerNodes) {
-    tracing::info!("Shutting down narwhal");
+    // Shuts down whole Narwhal (primary & worker(s)) and waits until nodes
+    // have shutdown.
+    pub async fn shutdown(&self) {
+        let mut running = self.running.lock().await;
 
-    primary_node.shutdown().await;
-    worker_nodes.shutdown().await;
+        let now = Instant::now();
+        tracing::info!("Shutting down Narwhal");
 
-    tracing::info!("Narwhal shutdown is complete");
-    // TODO: clean up previous storage disk as is no longer needed
-}
+        self.primary_node.shutdown().await;
+        self.worker_nodes.shutdown().await;
```

Comment:
Because of all these async calls, I don't think this will cause us to block until shutdown. I am not convinced this change overall is doing blocking shutdown. The approach we had before of waiting on the join handle was closer to a blocking shutdown, although we didn't signal back to sui when the join was complete.

Reply:
@laura-makdah actually because both the `shutdown` methods on the primary & worker nodes are waiting on the join handles, we will eventually block until the shutdown on the nodes have completed. For example, this is the implementation of the `shutdown` method for the primary node https://github.com/MystenLabs/sui/blob/88749d96a751909c58e870ae3a42eba3eaac1602/narwhal/node/src/primary_node.rs#L121